cunumeric/logic.py

```diff
@@ -0,0 +1,281 @@
+    """
+    x = convert_to_cunumeric_ndarray(x)
+    if x.dtype.kind != "c":
+        return full(x.shape, False, dtype=bool)
+    else:
+        return x.imag != 0
+
+
+def iscomplexobj(x):
+    """
+
+    Check for a complex type or an array of complex numbers.
+
+    The type of the input is checked, not the value. Even if the input
+    has an imaginary part equal to zero, `iscomplexobj` evaluates to True.
+
+    Parameters
+    ----------
+    x : any
+        The input can be of any type and shape.
+
+    Returns
+    -------
+    iscomplexobj : bool
+        The return value, True if `x` is of a complex type or has at least
+        one complex element.
+
+    See Also
+    --------
+    numpy.iscomplexobj
+
+    Availability
+    --------
+    Single CPU
+    """
+    if isinstance(x, ndarray):
+        return x.dtype.kind == "c"
+    else:
+        return np.iscomplexobj(x)
+
+
+def isreal(x):
+    """
+
+    Returns a bool array, where True if input element is real.
+
+    If element has complex type with zero complex part, the return value
+    for that element is True.
+
+    Parameters
+    ----------
+    x : array_like
+        Input array.
+
+    Returns
+    -------
+    out : ndarray, bool
+        Boolean array of same shape as `x`.
+
+
+    See Also
+    --------
+    numpy.isreal
+
+    Availability
+    --------
+    Multiple GPUs, Multiple CPUs
+
+    """
+    x = convert_to_cunumeric_ndarray(x)
+    if x.dtype.kind != "c":
+        return full(x.shape, True, dtype=bool)
+    else:
+        return x.imag == 0
+
+
+def isrealobj(x):
+    """
+
+    Return True if x is a not complex type or an array of complex numbers.
+
+    The type of the input is checked, not the value. So even if the input
+    has an imaginary part equal to zero, `isrealobj` evaluates to False
+    if the data type is complex.
+
+    Parameters
+    ----------
+    x : any
+        The input can be of any type and shape.
+
+    Returns
+    -------
+    y : bool
+        The return value, False if `x` is of a complex type.
+
+    See Also
+    --------
+    numpy.isrealobj
+
+    Availability
+    --------
+    Single CPU
+    """
+    return not iscomplexobj(x)
+
+
+def isscalar(x):
+    """
+
+    Returns True if the type of `element` is a scalar type.
+
+    Parameters
+    ----------
+    element : any
+        Input argument, can be of any type and shape.
+
+    Returns
+    -------
+    val : bool
+        True if `element` is a scalar type, False if it is not.
+
+    See Also
+    --------
+    numpy.isscalar
+
+    Notes
+    -----
+    This function falls back to NumPy for all object types but cuNumeric's
+    ndarray, which always returns `False`.
+
+    Availability
+    --------
+    Single CPU
+    """
+
+    # Since the input can be any value, we can't just convert it to cunumeric
+    # ndarray. Instead we check if the input is cunumeric ndarray and, if not,
+    # fall back to Numpy
+    if isinstance(x, ndarray):
+        return False
```

Comment:
I don't think this check is sufficient. It categorizes the results of full-array reductions as non-scalars:
```
print(np.isscalar(np.arange(3).sum())) # True
print(cn.isscalar(cn.arange(3).sum())) # False
```

Reply:
I disagree. It's cuNumeric's "fault" that it returns a singleton array where NumPy returns a Python scalar, and we shouldn't mask it by changing the notion of scalar but rather state it as one of the differences between cuNumeric and NumPy.